crates/holochain_state/src/mutations.rs

```diff
@@ -1154,3 +1154,273 @@ pub fn remove_countersigning_session(
+/// - If the op is not found in the cache, it returns early without error.
+/// - If the op is found in the cache, it constructs a `DhtOp` and inserts it into the DHT database
+///   using the same permit.
+///
+/// This function will fail on database errors or serialization errors but succeed if there was
+/// nothing to be done.
+pub async fn copy_cached_op_to_dht(
+    dht: DbWrite<DbKindDht>,
+    cache: DbRead<DbKindCache>,
+    action_hash: ActionHash,
+    chain_op_type: ChainOpType,
+) -> StateMutationResult<()> {
+    let dht_permit = dht.acquire_write_permit().await?;
+
+    let (exists_in_dht, dht_permit) = dht
+        .write_async_with_permit(dht_permit, {
+            let action_hash = action_hash.clone();
+            move |txn| -> StateMutationResult<bool> {
+                let exists = txn.query_row(
+                    "SELECT EXISTS (SELECT 1 FROM DhtOp WHERE action_hash = ? AND type = ?)",
+                    params![action_hash, chain_op_type],
+                    |row| row.get::<_, bool>(0),
+                )?;
+
+                Ok(exists)
+            }
+        })
+        .await?;
+
+    // Nothing further to do, a DhtOp with the same action_hash and type already exists
+    if exists_in_dht {
+        return Ok(());
+    }
+
+    let maybe_action_entry = cache
+        .read_async(
+            move |txn| -> StateMutationResult<Option<(SignedAction, Option<Entry>)>> {
+                let mut stmt = txn.prepare(
+                    r#"
+                        SELECT
+                          Action.blob AS action_blob,
+                          Entry.blob AS entry_blob
+                        FROM DhtOp
+                        JOIN Action ON DhtOp.action_hash = Action.hash
+                        LEFT JOIN Entry ON Action.entry_hash = Entry.hash
+                        WHERE
+                          DhtOp.type == :dht_op_type
+                          AND DhtOp.action_hash = :action_hash
+                        "#
+                )?;
+
+                let maybe_action_entry = stmt
+                    .query_row(
+                        named_params! {
+                            ":dht_op_type": chain_op_type,
+                            ":action_hash": action_hash,
+                        },
+                        |row: &Row| -> rusqlite::Result<(Vec<u8>, Option<Vec<u8>>)> {
+                            let action_blob = row.get::<_, Vec<u8>>(0)?;
+
+                            let entry_blob = row
+                                .get::<_, Option<Vec<u8>>>(1)?;
+
+                            Ok((action_blob, entry_blob))
+                        },
+                    )
+                    .optional()?.map(| (action_blob, entry_blob): (Vec<u8>, Option<Vec<u8>>) | -> StateMutationResult < (SignedAction, Option < Entry >) > {
+                    println!("Handling blobs: {:?} and {:?}", action_blob, entry_blob);
```

Comment:
```suggestion
```

Reply:
You don't like reading binary? It soothes me...

Yep this was a debugging change, removed